doc/design/draft/named-segments.md

```diff
@@ -0,0 +1,219 @@
+
+Currently initialization of the runtime happens in the constructor, and any errors that occur result in a contract violation and program termination. 
+
+While it is not directly in scope of this design, it would be beneficial to refactor the Runtime implementations to use the builder pattern such that any errors may be handled through returning errors types.
+
+Tying this together with requesting segments, this could look something like:
+```
+class DefaultRuntimeBuilder
+{
+  ...
+  IOX_BUILDER_PARAMETER(MappedSegments, premappedSegments, MappedSegments::ReadAndWrite)
+  IOX_BUILDER_PARAMETER(vector<ShmName_t, MAX_SHM_SEGMENTS>, shmSegmentNames, {});
+  IOX_BUILDER_PARAMETER(UnmappedSegmentBehavior, unmappedSegmentBehavior, UnmappedSegmentBehavior::LoadOnDemand);
+  ...
+  public:
+    expected<DefaultRuntime, RuntimeBuilderError> create() noexcept;
+```
+
+The `create` method of this builder would then return a custom error if:
+  * A segment is requested which does not exist
+  * A segment is requested which the current process does not have access to
+  * A segment is requested which has already been captured by the `premappedSegments` option
+
+Additionally, when a producer is created, the `unamppedSegmentBehavior` option would determine whether or not we fail or try to map the newly requested segment on the fly.
+
+### Segment Matching Algorithm
+
+Ensuring backwards compatibility means carefully crafting how we register segments. The following pseudocode demonstrates how this should work in different scenarios
+
+#### Producers requesting write access (Publisher, Client Request, Server Response)
+
+In RouDi, given:
+* `userGroups` - A list of POSIX user groups a publishing process belongs to called 
+* `segmentName` - The (possibly empty) name of a shared memory segment specified in the publisher options call
+* `mappedSegments` - A list of shared memory segments
+
+1. If `segmentName` is not empty
+    1. Find the segment in `mappedSegments` matching the `segmentName`
+    2. If the containing process has write access to the segment via one of its `userGroups`, return the corresponding segment information
+    3. Otherwise return an error indicating the publisher does not have write access to the requested segment
+2. If it is empty:
+    1. Iterate over all process `userGroups`
+    2. Determine if the user group name matches the name of one of the segments
+    3. If it does:
+        1. If a match has already been previously made, return an error indicating that the publisher must only have write access to one segment
+        2. If not, verify the process has access to the segment and record the match
+    4. At the end of iteration, if a matching segment has been found, return the segment information
+    5. Otherwise return an error indicating that no matching segment has been found
```

Comment:
Yeah... The reasons I mention RouDi doing this is because the logic is already there today: https://github.com/eclipse-iceoryx/iceoryx/blob/master/iceoryx_posh/source/roudi/process_manager.cpp#L493

The runtime sends an IPC request with publisher info and it is RouDi that checks the segment manager and figures out where that publisher should publish. 

It is an interesting thought to move some of this to the Runtime... I think you'd need to expose that API through the `SharedMemoryUser` class. I'm not sure that the runtime could go so far as to acquire the port data itself (though this moves more towards the decentralized architecture of iox2 and is a nice thought). It could however identify the segment it wishes to publish into and make a more specific request to roudi along the lines of "hey, I want a publisher port and I'm going to publish into THIS segment. Feel free to check on your end that I'm able to, though I already know I'm able to".